File: sdk-java/src/test/java/io/superdurable/dex/integ/StateOptionsTest.java
/*
 * Portions of this file are derived from indeedeng/iwf-java-sdk.
 * Those portions are licensed under the Apache License, Version 2.0.
 * See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
 *
 * Modifications Copyright (c) 2026 Super Durable, Inc.
 *
 * Modifications are licensed under the Super Durable Source License 1.0.
 * Third-Party Materials remain under the Apache License, Version 2.0.
 * See LICENSE and LEGACY_NOTICES.md.
 */

package io.superdurable.dex.integ;

import io.superdurable.dex.Client;
import io.superdurable.dex.testing.DexDevTestEnvironment;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import java.nio.file.Path;
import java.time.Duration;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.assertEquals;

@Tag("dex-dev")
public final class StateOptionsTest {
    private static final StateOptionsWorkflow WORKFLOW = new StateOptionsWorkflow();
    private static final StateOptionsLockingWorkflow LOCKING_WORKFLOW =
            new StateOptionsLockingWorkflow();

    @TempDir
    Path cacheDirectory;

    @Test
    void testStateOptionsWorkflow() throws Exception {
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                WORKFLOW)) {
            final String flowId = "state-options-" + UUID.randomUUID();
            environment.client().startFlow(WORKFLOW, flowId, null);
            assertEquals("success", environment.client().waitForFlow(
                    flowId,
                    String.class,
                    Duration.ofSeconds(30)));
        }
    }

    @Test
    void testWaitForAndExecuteLocksSerializeParallelSteps() throws Exception {
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                LOCKING_WORKFLOW)) {
            final String flowId = "state-options-locks-" + UUID.randomUUID();
            final int parallelism = 20;
            environment.client().startFlow(LOCKING_WORKFLOW, flowId, parallelism);
            assertEquals("20:20", environment.client().waitForFlow(
                    flowId,
                    String.class,
                    Duration.ofSeconds(30)));
            assertEquals(
                    parallelism,
                    environment.client().getAttribute(flowId, LOCKING_WORKFLOW.waitForCount));
            assertEquals(
                    parallelism,
                    environment.client().getAttribute(flowId, LOCKING_WORKFLOW.executeCount));
        }
    }

    void compileStepLocks(final Client client) {
        client.startFlow(LOCKING_WORKFLOW, "state-options-locks", 10);
        final String output = client.waitForFlow("state-options-locks", String.class);
        consume(output);
    }

    private static void consume(final Object value) {
    }
}

File: sdk-rust/crates/dex-sdk/tests/integ/state_options_test.rs
// Portions of this file are derived from indeedeng/iwf-java-sdk.
// Those portions are licensed under the Apache License, Version 2.0.
// See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
//
// Modifications Copyright (c) 2026 Super Durable, Inc.
//
// Modifications are licensed under the Super Durable Source License 1.0.
// Third-Party Materials remain under the Apache License, Version 2.0.
// See LICENSE and LEGACY_NOTICES.md.

use std::time::Duration;

use dex_sdk::{Client, Registry, SdkResult};

use crate::state_options_locking_workflow::{
    EXECUTE_COUNT, StateOptionsLockingWorkflow, WAIT_FOR_COUNT,
};
use crate::state_options_workflow::StateOptionsWorkflow;
use crate::support::{DexDevTestEnvironment, flow_id};

#[test]
#[ignore = "requires dexcli dev"]
fn test_state_options_workflow() {
    let environment =
        DexDevTestEnvironment::start(Registry::new().register(StateOptionsWorkflow::new()));
    let workflow = StateOptionsWorkflow::new();
    let flow_id = flow_id("state-options");
    environment
        .client
        .start_flow(&workflow, &flow_id, ())
        .expect("start state-options Flow");
    assert_eq!(
        "success",
        environment
            .client
            .wait_for_flow_with_timeout(&flow_id, Duration::from_secs(30))
            .and_then(|result| result.single_output::<String>())
            .expect("complete state-options Flow")
    );
}

#[test]
#[ignore = "requires dexcli dev"]
fn test_wait_for_and_execute_locks_serialize_parallel_steps() {
    let environment =
        DexDevTestEnvironment::start(Registry::new().register(StateOptionsLockingWorkflow::new()));
    let workflow = StateOptionsLockingWorkflow::new();
    let flow_id = flow_id("state-options-locks");
    let parallelism = 20;
    environment
        .client
        .start_flow(&workflow, &flow_id, parallelism)
        .expect("start step-locking Flow");
    assert_eq!(
        "20:20",
        environment
            .client
            .wait_for_flow_with_timeout(&flow_id, Duration::from_secs(30))
            .and_then(|result| result.single_output::<String>())
            .expect("complete step-locking Flow")
    );
    assert_eq!(
        Some(parallelism),
        environment
            .client
            .get_attribute(&flow_id, &WAIT_FOR_COUNT)
            .expect("read waitFor count")
    );
    assert_eq!(
        Some(parallelism),
        environment
            .client
            .get_attribute(&flow_id, &EXECUTE_COUNT)
            .expect("read execute count")
    );
}

#[allow(dead_code)]
fn compile_step_locks(client: &Client) -> SdkResult<()> {
    let workflow = StateOptionsLockingWorkflow::new();
    client.start_flow(&workflow, "state-options-locks", 10)?;
    let output: String = client
        .wait_for_flow("state-options-locks")?
        .single_output()?;
    drop(output);
    Ok(())
}
